lib/daru/io/importers/csv.rb

```diff
@@ -77,33 +85,59 @@ def initialize(path, headers: nil, col_sep: ',', converters: :numeric,
         def call
           # Preprocess headers for detecting and correcting repetition in
           # case the :headers option is not specified.
+
+          @file_string = process_compression
+
           hsh =
             if @headers
               hash_with_headers
             else
               hash_without_headers.tap { |hash| @daru_options[:order] = hash.keys }
             end
+
           Daru::DataFrame.new(hsh,@daru_options)
         end
 
         private
 
+        def compression?(algorithm, *formats)
+          return true if @compression == algorithm
+          formats.each { |f| return true if @path.end_with?(f) }
+          false
+        end
+
         def hash_with_headers
           ::CSV
-            .parse(open(@path), @options)
+            .parse(@file_string, @options)
             .tap { |c| yield c if block_given? }
-            .by_col.map { |col_name, values| [col_name, values] }.to_h
+            .by_col
+            .map do |col_name, values|
+              next [col_name, []] if values.nil?
+              [col_name, values[@skiprows..-1]]
```

Comment:
TBH, I believe that `return ... if` and `next .... if` is a last resort (when clear statement of entire algo is impossible, and you need to split it in steps). BUT!!!! If it looks cleaner for you and it is mindful solution -- let it stay. I want to share my experience, not to force my personal styles :)